// hbase-rpc-client/lib/data-input-stream.js
// Generated by CoffeeScript 1.12.7
(function() {
  var ByteBuffer, DataInputStream, EventEmitter, Readable,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Readable = require('readable-stream').Readable;

  ByteBuffer = require('bytebuffer');

  EventEmitter = require('events').EventEmitter;

  module.exports = DataInputStream = (function(superClass) {
    extend(DataInputStream, superClass);

    function DataInputStream(io) {
      this.processMessage = bind(this.processMessage, this);
      this.processData = bind(this.processData, this);
      DataInputStream.__super__.constructor.call(this);
      this["in"] = io;
      if (typeof io.read !== 'function') {
        this["in"] = new Readable();
        this["in"].wrap(io);
      }
      this.bytearr = new Buffer(80);
      this.buffer = new Buffer(0);
      this.awaitBytes = 0;
      this["in"].on('data', this.processData);
    }

    DataInputStream.prototype.processData = function(data) {
      var message;
      if (!data) {
        data = new Buffer(0);
      }
      this.buffer = Buffer.concat([this.buffer, data]);
      if (this.awaitBytes === 0 && this.buffer.length < 4) {
        return;
      }
      if (!this.awaitBytes) {
        this.awaitBytes = this.buffer.readUInt32BE(0);
        this.buffer = this.buffer.slice(4);
      }
      if (this.awaitBytes && this.awaitBytes > this.buffer.length) {
        return;
      }
      message = this.buffer.slice(0, this.awaitBytes);
      this.buffer = this.buffer.slice(this.awaitBytes);
      this.awaitBytes = 0;
      this.processMessage(message);
      if (this.buffer.length > 0) {
        return this.processData();
      }
    };

    DataInputStream.prototype.processMessage = function(message) {
      var messages, payload, readDelimited;
      payload = ByteBuffer.wrap(message);
      readDelimited = function() {
        var header, headerLen;
        headerLen = payload.readVarint32();
        header = payload.slice(payload.offset, payload.offset + headerLen);
        payload.offset += headerLen;
        return header.toBuffer();
      };
      messages = [];
      while (payload.remaining()) {
        messages.push(readDelimited());
      }
      return this.emit('messages', messages);
    };

    return DataInputStream;

  })(EventEmitter);

}).call(this);
